stakingModel/contracts/InterestDistribution.sol

```diff
@@ -0,0 +1,193 @@
+pragma solidity 0.5.4;
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+library InterestDistribution {
+    using SafeMath for uint256;
+
+    /**
+     * @dev Structure to store Interest details.
+     * It contains total amount of tokens staked and total amount of interest generated.
+     */    
+    struct InterestData {
+      uint grandTotalStaked;
+      uint interestAccrued;
+      uint accumulatedYieldPerToken;
+      mapping(address => Staker) stakers;
+    }
+
+    /**
+     * @dev Structure to store staking details.
+     * It contains amount of tokens staked and blocknumber at which last staked.
+     */
+    struct Staker {
+        uint256 stakedToken;
+        uint256 weightedStake; // stored with 2 precision points ie., 0.82 => 82
+        uint256 lastStake;
+        uint256 withdrawnToDate;
+        uint256 avgYieldRatePerToken;
+    }
+
+    // 10^18
+    uint256 constant DECIMAL1e18 = 10 ** 18;
+
+    /**
+      * @dev Updates InterestData and Staker data while staking.
+      * 
+      * @param _interestData           Interest data
+      * @param _staker                 Staker's address
+      * @param _stake                  Amount of stake
+      * @param _donationPer            Percentage will to donate.
+      * @param _dailyIntrest           Withdrawn interest by individual staker so far.
+      * @param _grandTotalStaked       Average yielding rate per token
+      *
+    */
+    function stakeCalculation(
+      InterestData storage _interestData, 
+      address _staker,
+      uint256 _stake, 
+      uint256 _donationPer,
+      uint256 _dailyIntrest,
+      uint256 _grandTotalStaked
+      ) 
+    internal 
+    {
+      _interestData.grandTotalStaked = _interestData.grandTotalStaked.add(_stake);
+      uint currentStake = _interestData.stakers[_staker].stakedToken;
+      Staker storage _stakerData = _interestData.stakers[_staker];
+      _stakerData.stakedToken = currentStake.add(_stake);
+      _stakerData.weightedStake = (_stakerData.weightedStake.mul(currentStake).add(_stake.mul(uint(100).sub(_donationPer)))).div(_stakerData.stakedToken);
+      _stakerData.lastStake = block.number;
+      uint _accumulatedYieldPerToken = getAccumulatedYieldPerToken(_dailyIntrest, _grandTotalStaked);
+      _stakerData.avgYieldRatePerToken = _stakerData.avgYieldRatePerToken.add(getAvgYieldRatePerToken(_accumulatedYieldPerToken, _stake, _donationPer, _stakerData.stakedToken));
+    }
+
+    /**
+      * @dev Updates InterestData and Staker data while withdrawing stake.
+      * 
+      * @param _interestData           Interest data
+      * @param _staker                 Staker address
+      * @param _amount                 Amount of stake to withdraw
+      *
+    */
+    function withdrawStakeAndInterest(
```

Comment:
you didnt withdraw the interest only the stake

Reply:
done.